In [1]:
import pandas as pd

In [2]:
df_fusionné = pd.read_csv('df_fusionné.csv')

In [3]:
df_fusionné_binaire = df_fusionné.sample(frac=0.2)
df_fusionné_binaire.iloc[:, 0] = df_fusionné_binaire.iloc[:, 0].apply(lambda x: 0 if x == 'N' else 1)

In [4]:
# Count the occurrences of each class
counts = df_fusionné_binaire.iloc[:, 0].value_counts()

# Find the number of instances in the minority class
min_count = counts.min()

# Create balanced DataFrame through undersampling
df_balanced_under = pd.concat([
    df_fusionné_binaire[df_fusionné_binaire.iloc[:, 0] == 0].sample(min_count),
    df_fusionné_binaire[df_fusionné_binaire.iloc[:, 0] == 1].sample(min_count)
])

# Shuffle the DataFrame to mix the classes
df_balanced_under = df_balanced_under.sample(frac=1)


In [5]:
# Count occurrences of 0 and 1 in the first column
value_counts = df_balanced_under.iloc[:, 0].value_counts()

# Display the counts
print(value_counts)

ColumnName
1    7552
0    7552
Name: count, dtype: int64


In [6]:
df_balanced_under

ColumnName         0         1         2         3         4         5  \
33889           1  0.534183  0.532404  0.530472  0.528452  0.528661  0.529608   
11513           1  0.495040  0.495883  0.496152  0.495715  0.496181  0.496769   
167175          1  0.517586  0.518428  0.519394  0.520493  0.521597  0.522739   
123956          1  0.501004  0.524533  0.546563  0.568909  0.589660  0.605683   
10961           1  0.698745  0.692443  0.708779  0.685442  0.692073  0.699626   
...           ...       ...       ...       ...       ...       ...       ...   
141355          0  0.512862  0.512748  0.512619  0.512476  0.512336  0.512191   
70712           0  0.396079  0.393459  0.388059  0.392337  0.390520  0.388839   
191698          1  0.441102  0.440656  0.440203  0.439760  0.439337  0.438913   
18587           1  0.523958  0.525773  0.527352  0.528821  0.530613  0.532341   
13644           0  0.503153  0.504026  0.508054  0.503114  0.504208  0.505181   

               6         7         8  ...       710       711       712  \
33889   0.530814  0.532115  0.532327  ...  0.556601  0.576580  0.572279   
11513   0.495641  0.494058  0.494252  ...  0.250378  0.239728  0.234463   
167175  0.523979  0.525248  0.526422  ...  0.518043  0.518102  0.517992   
123956  0.627932  0.645150  0.626935  ...  0.482364  0.483960  0.483293   
10961   0.690838  0.677497  0.675879  ...  0.502441  0.505818  0.499260   
...          ...       ...       ...  ...       ...       ...       ...   
141355  0.512038  0.511879  0.511714  ...  0.513092  0.513062  0.513047   
70712   0.391488  0.395254  0.394994  ...  0.430992  0.429413  0.432892   
191698  0.438511  0.438099  0.437577  ...  0.445797  0.445559  0.445211   
18587   0.534081  0.535594  0.536022  ...  0.507455  0.506640  0.507495   
13644   0.502315  0.498273  0.496763  ...  0.463075  0.464947  0.463246   

             713       714       715       716       717       718       719  
33889   0.559702  0.542205  0.523533  0.521134  0.524261  0.528782  0.534296  
11513   0.230628  0.225351  0.224546  0.207494  0.208153  0.225726  0.207573  
167175  0.517814  0.517722  0.517645  0.517450  0.517301  0.517110  0.517039  
123956  0.481984  0.482905  0.484445  0.484032  0.484953  0.483466  0.483898  
10961   0.489663  0.493678  0.504173  0.494113  0.494219  0.506399  0.493096  
...          ...       ...       ...       ...       ...       ...       ...  
141355  0.513039  0.513022  0.513003  0.512995  0.512981  0.512970  0.512947  
70712   0.437377  0.441990  0.446018  0.442883  0.433962  0.435085  0.445113  
191698  0.444794  0.444355  0.443879  0.443344  0.442787  0.442191  0.441592  
18587   0.509105  0.510989  0.513165  0.515260  0.517371  0.519687  0.522001  
13644   0.460902  0.458563  0.456666  0.460249  0.468640  0.467697  0.458149  

[15104 rows x 721 columns]

In [7]:
import pandas as pd
from sklearn.model_selection import train_test_split

X = df_balanced_under.iloc[:, 1:]  
y = df_balanced_under.iloc[:, 0] 

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [8]:
print(y_train.unique())

[1 0]


In [9]:
y_train = y_train.astype(int)
y_test = y_test.astype(int)

In [10]:
import tensorflow as tf
from keras.models import Sequential
from keras.layers import Conv1D, BatchNormalization, MaxPooling1D, Dropout, Flatten, Dense
from keras.optimizers import Adam

Model = Sequential()

Model.add(Conv1D(512, 10, strides=1, padding="same", activation="relu", input_shape=(720, 1)))
Model.add(BatchNormalization())
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(256, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.2))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(1024, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.2))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(128, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.2))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(64, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.2))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(32, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.2))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Flatten())
Model.add(Dense(1024, activation='relu'))
Model.add(Dropout(0.2))

Model.add(Dense(512, activation='relu'))
Model.add(Dropout(0.2))

Model.add(Dense(64, activation='relu'))
Model.add(Dropout(0.2))

Model.add(Dense(units=1, activation='sigmoid'))  # Use 'sigmoid' for binary classification


/Users/arthurdesmazures/venv-metal/lib/python3.10/site-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
2024-11-07 16:51:44.855134: I metal_plugin/src/device/metal_device.cc:1154] Metal device set to: Apple M3 Pro
2024-11-07 16:51:44.855159: I metal_plugin/src/device/metal_device.cc:296] systemMemory: 18.00 GB
2024-11-07 16:51:44.855164: I metal_plugin/src/device/metal_device.cc:313] maxCacheSize: 6.00 GB
2024-11-07 16:51:44.855334: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:305] Could not identify NUMA node of platform GPU ID 0, defaulting to 0. Your kernel may not have been built with NUMA support.
2024-11-07 16:51:44.855343: I tensorflow/core/common_runtime/pluggable_device/pluggable_devi

In [11]:
print(Model.summary())

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv1d (Conv1D)                 │ (None, 720, 512)       │         5,632 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization             │ (None, 720, 512)       │         2,048 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d (MaxPooling1D)    │ (None, 360, 512)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_1 (Conv1D)               │ (None, 360, 256)       │     1,310,976 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 360, 256)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_1 (MaxPooling1D)  │ (None, 180, 256)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_2 (Conv1D)               │ (None, 180, 1024)      │     2,622,464 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 180, 1024)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_2 (MaxPooling1D)  │ (None, 90, 1024)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_3 (Conv1D)               │ (None, 90, 128)        │     1,310,848 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_2 (Dropout)             │ (None, 90, 128)        │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_3 (MaxPooling1D)  │ (None, 45, 128)        │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_4 (Conv1D)               │ (None, 45, 64)         │        81,984 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_3 (Dropout)             │ (None, 45, 64)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_4 (MaxPooling1D)  │ (None, 23, 64)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_5 (Conv1D)               │ (None, 23, 32)         │        20,512 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_4 (Dropout)             │ (None, 23, 32)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_5 (MaxPooling1D)  │ (None, 12, 32)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 384)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 1024)           │       394,240 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_5 (Dropout)             │ (None, 1024)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 512)            │       524,800 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_6 (Dropout)             │ (None, 512)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 64)             │        32,832 │
├─────────────────────────────────┼────────────────────────┼─────────────

 Total params: 6,306,401 (24.06 MB)

 Trainable params: 6,305,377 (24.05 MB)

 Non-trainable params: 1,024 (4.00 KB)

None


In [12]:
# Define callbacks
Early_Stopper = tf.keras.callbacks.EarlyStopping(monitor="loss", patience=3, mode="min")


In [13]:
# Compile the model
Model.compile(optimizer=Adam(), loss='binary_crossentropy', metrics=['accuracy'])

In [14]:
Conv1D_Model = Model.fit(X_train, y_train, epochs=70, validation_data=(X_test, y_test), callbacks=[Early_Stopper])

Epoch 1/70


2024-11-07 16:51:45.811329: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:117] Plugin optimizer for device_type GPU is enabled.


378/378 ━━━━━━━━━━━━━━━━━━━━ 53s 135ms/step - accuracy: 0.6818 - loss: 0.6004 - val_accuracy: 0.5717 - val_loss: 0.7488
Epoch 2/70
378/378 ━━━━━━━━━━━━━━━━━━━━ 49s 130ms/step - accuracy: 0.7922 - loss: 0.4570 - val_accuracy: 0.6425 - val_loss: 0.6671
Epoch 3/70
378/378 ━━━━━━━━━━━━━━━━━━━━ 50s 133ms/step - accuracy: 0.8209 - loss: 0.4137 - val_accuracy: 0.7259 - val_loss: 0.5154
Epoch 4/70
378/378 ━━━━━━━━━━━━━━━━━━━━ 50s 133ms/step - accuracy: 0.8293 - loss: 0.3988 - val_accuracy: 0.7630 - val_loss: 0.5041
Epoch 5/70
378/378 ━━━━━━━━━━━━━━━━━━━━ 50s 133ms/step - accuracy: 0.8521 - loss: 0.3547 - val_accuracy: 0.6438 - val_loss: 1.2364
Epoch 6/70
378/378 ━━━━━━━━━━━━━━━━━━━━ 50s 133ms/step - accuracy: 0.8576 - loss: 0.3422 - val_accuracy: 0.7699 - val_loss: 0.4798
Epoch 7/70
378/378 ━━━━━━━━━━━━━━━━━━━━ 50s 133ms/step - accuracy: 0.8667 - loss: 0.3332 - val_accuracy: 0.5766 - val_loss: 1.0698
Epoch 8/70
378/378 ━━━━━━━━━━━━━━━━━━━━ 50s 132ms/step - accuracy: 0.8690 - loss: 0.3185 - val

In [15]:
Model.save('Conv1D_Model.h5')  # Saves the model to a HDF5 file

In [16]:
Model_Results = Model.evaluate(X_test,y_test)
print("LOSS:  " + "%.4f" % Model_Results[0])
print("ACCURACY:  " + "%.4f" % Model_Results[1])

95/95 ━━━━━━━━━━━━━━━━━━━━ 2s 26ms/step - accuracy: 0.8922 - loss: 0.2839
LOSS:  0.2846
ACCURACY:  0.8908


In [17]:
# Assuming your model is named 'Model'
y_pred_probs = Model.predict(X_test)
y_pred = (y_pred_probs > 0.5).astype("int32")  # Convert probabilities to binary labels


95/95 ━━━━━━━━━━━━━━━━━━━━ 2s 24ms/step


In [18]:
from sklearn.metrics import confusion_matrix, precision_score, recall_score, f1_score, roc_auc_score, roc_curve, log_loss

# Confusion Matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:\n", conf_matrix)

# Precision
precision = precision_score(y_test, y_pred)
print(f"Precision: {precision:.2f}")

# Recall
recall = recall_score(y_test, y_pred)
print(f"Recall: {recall:.2f}")

# F1 Score
f1 = f1_score(y_test, y_pred)
print(f"F1 Score: {f1:.2f}")

# ROC-AUC Score
roc_auc = roc_auc_score(y_test, y_pred_probs)
print(f"ROC AUC Score: {roc_auc:.2f}")

# Now, you can calculate the log loss
logloss = log_loss(y_test, y_pred_probs)
print(f"Log Loss: {logloss:.2f}")

Confusion Matrix:
 [[1332  173]
 [ 157 1359]]
Precision: 0.89
Recall: 0.90
F1 Score: 0.89
ROC AUC Score: 0.96
Log Loss: 0.28


Confusion Matrix:
 [[1359  157]
 [ 167 1354]]
Precision: 0.90
Recall: 0.89
F1 Score: 0.89
ROC AUC Score: 0.96
Log Loss: 0.30
